crates/matrix-sdk/src/room_directory_search.rs

```diff
@@ -0,0 +1,370 @@
+    /// The room's avatar URL, if any.
+    pub avatar_url: Option<OwnedMxcUri>,
+    /// The room's join rule.
+    pub join_rule: PublicRoomJoinRule,
+    /// Whether can be previewed
+    pub is_world_readable: bool,
+    /// The number of members that have joined the room.
+    pub joined_members: u64,
+}
+
+impl From<ruma::directory::PublicRoomsChunk> for RoomDescription {
+    fn from(value: ruma::directory::PublicRoomsChunk) -> Self {
+        Self {
+            room_id: value.room_id,
+            name: value.name,
+            topic: value.topic,
+            alias: value.canonical_alias,
+            avatar_url: value.avatar_url,
+            join_rule: value.join_rule,
+            is_world_readable: value.world_readable,
+            joined_members: value.num_joined_members.into(),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct RoomDirectorySearch {
+    batch_size: u32,
+    filter: Option<String>,
+    next_token: Option<String>,
+    client: Client,
+    results: ObservableVector<RoomDescription>,
+    is_at_last_page: bool,
+}
+
+impl RoomDirectorySearch {
+    pub fn new(client: Client) -> Self {
+        Self {
+            batch_size: 0,
+            filter: None,
+            next_token: None,
+            client,
+            results: ObservableVector::new(),
+            is_at_last_page: false,
+        }
+    }
+
+    /// Starts a filtered search for the server
+    /// If the `filter` is not provided it will search for all the rooms
+    /// You can specify a `batch_size`` to control the number of rooms to fetch
+    /// per request
+    ///
+    /// This method will clear the current search results and start a new one
+    /// Should never be used concurrently with another `next_page` or a
+    /// `search`.
+    pub async fn search(&mut self, filter: Option<String>, batch_size: u32) -> Result<()> {
+        self.filter = filter;
+        self.batch_size = batch_size;
+        self.next_token = None;
+        self.results.clear();
+        self.is_at_last_page = false;
+        self.next_page().await
+    }
+
+    /// Asks the server for the next page of the current search
+    /// Should never be used concurrently with another `next_page` or a
+    /// `search`.
+    pub async fn next_page(&mut self) -> Result<()> {
+        if self.is_at_last_page {
+            return Ok(());
+        }
+        let mut filter = Filter::new();
+        filter.generic_search_term = self.filter.clone();
+
+        let mut request = PublicRoomsFilterRequest::new();
+        request.filter = filter;
+        request.limit = Some(self.batch_size.into());
+        request.since = self.next_token.clone();
+        let response = self.client.public_rooms_filtered(request).await?;
+        self.next_token = response.next_batch;
+        if self.next_token.is_none() {
+            self.is_at_last_page = true;
+        }
+        self.results.append(response.chunk.into_iter().map(Into::into).collect());
+        Ok(())
```

Comment:
The code needs to breath too.

```suggestion
        if self.is_at_last_page {
            return Ok(());
        }

        let mut filter = Filter::new();
        filter.generic_search_term = self.filter.clone();

        let mut request = PublicRoomsFilterRequest::new();
        request.filter = filter;
        request.limit = Some(self.batch_size.into());
        request.since = self.next_token.clone();

        let response = self.client.public_rooms_filtered(request).await?;
        self.next_token = response.next_batch;
        self.is_at_last_page = self.next_token.is_none();

        self.results.append(response.chunk.into_iter().map(Into::into).collect());

        Ok(())
```

Reply:
I also realize that `is_at_last_page` is useless and can be replaced by `next_token.is_none()`. It would reduce the size of the `RoomDirectorySearch` type, and will simplify the code a little bit. Less field is often better.

Reply:
Oh, so we would need to encore this as a new enum:

```rust
#[derive(Default)
enum NextToken {
    Current(String),
    End,
    #[default] // <- this is the default variant
    Uninitialized,
}

// if you want extra methods:
impl NextToken {
    fn is_current(&self) -> bool {
        matches!(self, Self::Current)
    }

    fn current(&self) -> Option<&str> {
        if let Self::Current(next_token) = &self {
            Some(next_token)
        } else {
            None
        }
    }

    fn is_end(&self) -> bool {
        matches!(self, Self::End)
    }
}
```

It's a bit more code for sure, but I think it conveys the intent more clearly.

Thoughts?